includes/sanitizers/class-amp-style-sanitizer.php

```diff
@@ -349,230 +431,741 @@ private function process_link_element( DOMElement $element ) {
+			if ( 'data:' === substr( $url_string, 0, 5 ) ) {
+				continue;
+			}
 
-		if ( ! empty( $properties ) ) {
-			$class_name = $this->generate_class_name( $properties );
-			$new_class  = trim( $class . ' ' . $class_name );
+			$parsed_url = wp_parse_url( $url_string );
+			if ( ! empty( $parsed_url['host'] ) || empty( $parsed_url['path'] ) || '/' === substr( $parsed_url['path'], 0, 1 ) ) {
+				continue;
+			}
 
-			$selector = '.' . $class_name;
-			$length   = strlen( sprintf( '%s { %s }', $selector, join( '; ', $properties ) . ';' ) );
+			$relative_url = preg_replace( '#^\./#', '', $url->getURL()->getString() );
 
-			if ( $this->current_custom_size + $length > $this->custom_max_size ) {
-				$this->remove_invalid_attribute( $element, 'style', array(
-					'message' => __( 'Too much CSS.', 'amp' ),
-				) );
-				return;
-			}
+			$real_url = $base_url . $relative_url;
+			do {
+				$real_url = preg_replace( '#[^/]+/../#', '', $real_url, -1, $count );
+			} while ( 0 !== $count );
 
-			$element->setAttribute( 'class', $new_class );
-			$this->styles[ $selector ] = $properties;
+			$url->getURL()->setString( $real_url );
 		}
-		$element->removeAttribute( 'style' );
 	}
 
 	/**
-	 * Sanitize and convert individual styles.
+	 * Process CSS rule set.
 	 *
-	 * @since 0.4
+	 * @since 1.0
+	 * @link https://www.ampproject.org/docs/design/responsive/style_pages#disallowed-styles
+	 * @link https://www.ampproject.org/docs/design/responsive/style_pages#restricted-styles
+	 *
+	 * @param RuleSet $ruleset  Ruleset.
+	 * @param CSSList $css_list CSS List.
+	 * @param array   $options  Options.
 	 *
-	 * @param string $css Style string.
-	 * @return array Style properties.
+	 * @return array Validation errors.
 	 */
-	private function process_style( $css ) {
+	private function process_css_declaration_block( RuleSet $ruleset, CSSList $css_list, $options ) {
+		$validation_errors = array();
+
+		// Remove disallowed properties.
+		if ( ! empty( $options['property_whitelist'] ) ) {
+			$properties = $ruleset->getRules();
+			foreach ( $properties as $property ) {
+				$vendorless_property_name = preg_replace( '/^-\w+-/', '', $property->getRule() );
+				if ( ! in_array( $vendorless_property_name, $options['property_whitelist'], true ) ) {
+					$validation_errors[] = array(
+						'code'           => 'illegal_css_property',
+						'property_name'  => $property->getRule(),
+						'property_value' => $property->getValue(),
+					);
+					$ruleset->removeRule( $property->getRule() );
+				}
+			}
+		} else {
+			foreach ( $options['property_blacklist'] as $illegal_property_name ) {
+				$properties = $ruleset->getRules( $illegal_property_name );
+				foreach ( $properties as $property ) {
+					$validation_errors[] = array(
+						'code'           => 'illegal_css_property',
+						'property_name'  => $property->getRule(),
+						'property_value' => $property->getValue(),
+					);
+					$ruleset->removeRule( $property->getRule() );
```

Comment:
Consider skipping the removal (or let it be configurable) since it could have unintended results, but since `-moz-binding` and `behavior` are the only blacklisted properties (for now), it may not be the end of the world if they are removed regardless.